### PR TITLE
Moved staging/src/k8s.io/client-go/tools/bootstrap to staging/src/k8s…

### LINKED ISSUE #64627: Find a new home for bootstrap tokens consts & utilities
**This is a CLEANUP/ REFACTOR **:
During 1.11 release, as part of https://github.com/kubernetes/kubernetes/pull/64408/commits/33f59e438e93b824492caff5d39fd143b43eac9a there were concerns raised by @lavalamp regarding whether the aforementioned code belongs there, or in a different location. 

from slack: 
```
Yeah, an "api" that's ONLY in the *go* client, under a random tools directory, is not really acceptable. Doesn't work for other languages, and it can't be enforced correctly, etc. If it's just a convention then some combination of: it shouldn't be called an api (?), should be documented such that you can repeat from other languages, should be implemented in a control plane component and not a client, should be promoted to an actual api.
And if it's specific to kubeadm, which it appears to be (?), then it shouldn't be in the general go client.
Also I'm very nervous about a security-critical feature being non obviously buried here
```

The purpose of this issue is to find that home in the 1.12 cycle. 

/kind cleanup 
/cc @kubernetes/sig-cluster-lifecycle-misc @kubernetes/sig-api-machinery-misc

### PR DESCRIPTION
**Pull request by yliaog**
**What this PR does / why we need it**:
Moved bootstrap out of client-go, where it does not belong to.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64627

**Special notes for your reviewer**:

**Dev-Release note**:

```dev-release-note
action required: client-go release will no longer have bootstrap (k8s.io/client-go/tools/bootstrap) related code. any reference to it will break. Please redirect all references to k8s.io/bootstrap instead.
```

### CODE REVIEW

File: hack/import-restrictions.yaml
Comment (sttts) [inline hack/import-restrictions.yaml]: this line can go.

But we need another similar block for the whole k8s.io/bootstrap directory.

**Comment by sttts** [comment]
Please open a PR against https://github.com/kubernetes/publishing-bot/blob/master/configs/kubernetes-rules-configmap.yaml with the new repo.

File: plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go
Comment (sttts) [inline plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go]: why is this copied?

Comment (yliaog) [reply plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go]: i'm not exactly sure why it was copied before. this PR moves the files from client-go to bootstrap, without touching the files' functions. i would leave it to the original author to fix the copy if necessary.

**Comment by yliaog** [comment]
https://github.com/kubernetes/publishing-bot/pull/89 for adding the bootstrap repo

File: hack/import-restrictions.yaml
Comment (yliaog) [reply hack/import-restrictions.yaml]: https://github.com/kubernetes/publishing-bot/pull/89

Comment (yliaog) [reply hack/import-restrictions.yaml]: https://github.com/kubernetes/kubernetes/pull/67356/commits/06a3a3bcd7f891d3074bcfd83e5215d99e068c11

File: staging/src/k8s.io/bootstrap/OWNERS
Comment (caesarxuchao) [inline staging/src/k8s.io/bootstrap/OWNERS]: I don't really know this code area, please remove me from approvers and reviewers. Thanks.

**Review by caesarxuchao — COMMENTED** [review]
/lgtm

File: plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go
Comment (sttts) [reply plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go]: @luxas @timstclair you probably know

**Comment by sttts** [comment]
/approve
/lgtm

/assign @lavalamp 

for toplevel approval.

**Comment by sttts** [comment]
@yliaog updated the release note. This is only relevant for client-go releases, not for the normal kube release notes.

**Comment by smarterclayton** [comment]
`bootstrap` is a pretty broad name, can you scope it to `cluster-bootstrap` or something more precise?

**Comment by yliaog** [comment]
it was originally named 'bootstrap', this PR is simply moving it around. Since it is going to be used in import in various places, a concise name is probably better. Instead of changing the name, what about adding more descriptions to the doc of the package to make it clear what it is meant to do?

File: staging/src/k8s.io/bootstrap/OWNERS
Comment (timothysc) [inline staging/src/k8s.io/bootstrap/OWNERS]: Please add sig-cluster-lifecycle folks to the approvers+reviewers list.

**Review by timothysc — COMMENTED** [review]
minor comment. 

/cc @luxas @fabriziopandini @liztio 

Thanks for doing this, it simplifies our side.

**Comment by lavalamp** [comment]
looks great, but I thought the name was longer than just k8s.io/bootstrap?

File: staging/src/k8s.io/bootstrap/OWNERS
Comment (yliaog) [reply staging/src/k8s.io/bootstrap/OWNERS]: added sig-cluster-lifecycle folks to the approvers+reviewers list

**Comment by yliaog** [comment]
also changed to k8s.io/cluster-bootstrap

**Comment by yliaog** [comment]
squashed the commits, PTAL

**Comment by sttts** [comment]
@yliaog please rebase

**Comment by dims** [comment]
/milestone v1.12 

(the associated issue is marked v1.12)

**Review by luxas — COMMENTED** [review]
/approve

But it needs a rebase...

File: staging/src/k8s.io/metrics/metrics.go
Comment (luxas) [reply staging/src/k8s.io/metrics/metrics.go]: yeah please remove

Comment (yliaog) [reply staging/src/k8s.io/metrics/metrics.go]: Yes, it is an accident. removed.

**Comment by yliaog** [comment]
rebased

**Comment by sttts** [comment]
Need a rebase, and tests are broken.

File: staging/src/k8s.io/cluster-bootstrap/OWNERS
Comment (sttts) [inline staging/src/k8s.io/cluster-bootstrap/OWNERS]: these are a lot of owners. Who is the real owner of this? Would expect more like 3 people.

Comment (timothysc) [reply staging/src/k8s.io/cluster-bootstrap/OWNERS]: sig-cluster-lifecycle

Comment (yliaog) [reply staging/src/k8s.io/cluster-bootstrap/OWNERS]: removed all others, replaced with sig-cluster-lifecycle-leads

**Review by luxas — APPROVED** [review]
/lgtm
/assign @jbeda @bgrant0607 @lavalamp 
for approval

**Comment by timothysc** [comment]
@kubernetes/sig-architecture-pr-reviews  
@kubernetes/sig-api-machinery-pr-reviews  

What's the plan here? Is this still blocked on final name, and location?

**Comment by yliaog** [comment]
No, it is not blocked there, but by the missing label "Will not auto merge because priority/critical-urgent is missing". I'm wondering if it's ok to add that label.

**Comment by liggitt** [comment]
This was discussed in https://groups.google.com/forum/m/?utm_medium=email&utm_source=footer#!topic/kubernetes-sig-architecture/d_E7JOyA4EM and unresolved. I think a whole repo for a file of constants and a file of helper functions is probably too granular. Would this make more sense under `k8s.io/kube-apiserver/<bootstrap-module>`

**Comment by yliaog** [comment]
@liggitt I think we discussed the proper location for these files, in particular in the comment (https://github.com/kubernetes/kubernetes/issues/64627#issuecomment-395905283) there is concern about placing them under *-apiserver-*

I'm not sure about whether there will be more developments going forward around it. If there is no immediate need for publishing it as a repo, one option I can think of is to keep it under staging without creating a repo for now.

**Comment by bgrant0607** [comment]
@liggitt @yliaog I agree with @lavalamp that clients should not depend on server code.

**Comment by bgrant0607** [comment]
I suggest moving ahead with the repo unless/until we find a better solution later.

**Comment by liggitt** [comment]
> @liggitt @yliaog I agree with @lavalamp that clients should not depend on server code

I agree they should not depend on server code. The consumers of cluster bootstrap token helpers seemed to align more closely with the ones that would be concerned with server *config*, which is intended to be a distinct package/module under k8s.io/kube-apiserver, not dependent on any server code.

**Comment by yliaog** [comment]
thanks liggitt, we can revisit it once we can find a better home for it.

/label priority/critical-urgent

**Comment by liggitt** [comment]
> we can revisit it once we can find a better home for it

can we? Isn't moving things into external repos explicitly so outside projects can take dependencies on them? The current readme doesn't make it clear what the compatibility guarantee around the repo is.

**Comment by spiffxp** [comment]
Please file an issue against kubernetes/org if you'd like the new repo to be created

**Comment by guineveresaenger** [comment]
@liggitt @yliaog see my comment[here](https://github.com/kubernetes/kubernetes/issues/64627#issuecomment-419511537)

Technically, this is not critical but since all the work is done if your conscience can handle applying the label, please do, since that's all this PR is hung on.

**Comment by liggitt** [comment]
Every addition of a staging repo involves additional work to create a new actual repo and configure the publishing bot to push changes out. 

@sttts @dims where is that work at for the already-created repos that are new in 1.12? Would adding another one be feasible?

**Comment by dims** [comment]
@liggitt if we can merge the code today and get the repo created as well, then we will have a better chance. @sttts is wrangling the bot. So we should be able to do one more.

**Comment by dims** [comment]
cc @cblecker

**Comment by dims** [comment]
@yliaog @liggitt if we want to do this, there's a bunch of things we need to do on Sep 10 since beta.2 is on next day Sept 11

**Comment by liggitt** [comment]
> @yliaog @liggitt if we want to do this, there's a bunch of things we need to do on Sep 10 since beta.2 is on next day Sept 11

I dont feel strongly either way.  I'll let the sig-cluster-lifecycle leads make the call since the new repo is scoped to their sig.

**Comment by yliaog** [comment]
I agree with @liggitt , I'll leave it to sig-cluster-lifecycle (@timothysc @luxas) to make the call

**Comment by dims** [comment]
/hold

It's too late now (cc @tpepper)

**Comment by tpepper** [comment]
I feel this is too late for v1.12.

**Comment by guineveresaenger** [comment]
/milestone clear

Apologies. Please revisit one master reopens.

**Comment by yliaog** [comment]
@timothysc @luxas it looks the lgtm label was removed, do you mind to take a look again?

**Review by timothysc — APPROVED** [review]
/lgtm 

/cc @fabriziopandini

**Comment by yliaog** [comment]
again, @timothysc @luxas it looks the lgtm label was removed, do you mind to take a look again?

**Pull request merged**
